### PR TITLE
159 implement remove/disconnect participants functionality

### PR DESCRIPTION
**Pull request by farmanahmed888**
This pull request adds the ability to remove participants from a conference call in the teacher webapp and backend API. It updates the UI to allow teachers to remove students, handles the removal process via API, and ensures that the participant state is correctly updated in real time via SSE events. The changes also include several backend bug fixes for route handler naming.

**Backend API improvements:**

* Fixed route handler naming in `conference.py` to correctly match their purpose (e.g., `remove_participant`, `pause_audio`, `resume_audio`) for clarity and correctness. [[1]](diffhunk://#diff-17184d52196cf20cd18856c0c3ce67c65db2edb6c6cc0545da1c0641416f2941L104-R104) [[2]](diffhunk://#diff-17184d52196cf20cd18856c0c3ce67c65db2edb6c6cc0545da1c0641416f2941L146-R146) [[3]](diffhunk://#diff-17184d52196cf20cd18856c0c3ce67c65db2edb6c6cc0545da1c0641416f2941L155-R155)

**Frontend feature: Remove participant**

* Added `removeParticipant` API endpoint and service method, allowing the frontend to request participant removal from a conference. [[1]](diffhunk://#diff-853971ee113eeda9a96d9ee097e224bfd66cf0f9a98df4bb7cff057dc8ef281eR33-R34) [[2]](diffhunk://#diff-0da0db9ad87dbd387912bd10416858eb70f8248f0845495dde4a503e1b0e7f51R149-R176) [[3]](diffhunk://#diff-3157ebb9fc36a49a9275ff7bf9ac4028dd3d9fd922ac03dfac5b4a049ab98334R14)
* Implemented participant removal UI: added "Remove" buttons to `ParticipantCard` for students, with confirmation dialogs and loading indicators. [[1]](diffhunk://#diff-0ac8b38eb227b3f2273c87ffc4122c56ae8c5dc8812c9914dfdea44b832a9e01R18-R29) [[2]](diffhunk://#diff-0ac8b38eb227b3f2273c87ffc4122c56ae8c5dc8812c9914dfdea44b832a9e01R122-R138)
* Updated participant list and details page to support removal, including tracking removal state and passing callbacks through the component tree. [[1]](diffhunk://#diff-3157ebb9fc36a49a9275ff7bf9ac4028dd3d9fd922ac03dfac5b4a049ab98334R41) [[2]](diffhunk://#diff-3157ebb9fc36a49a9275ff7bf9ac4028dd3d9fd922ac03dfac5b4a049ab98334R259-R297) [[3]](diffhunk://#diff-3157ebb9fc36a49a9275ff7bf9ac4028dd3d9fd922ac03dfac5b4a049ab98334R324) [[4]](diffhunk://#diff-3157ebb9fc36a49a9275ff7bf9ac4028dd3d9fd922ac03dfac5b4a049ab98334R348-R351) [[5]](diffhunk://#diff-e383c735a97293f754019abc3e8599b68378e65dd6b4884ebadc7d607459fb2aR12-R15) [[6]](diffhunk://#diff-e383c735a97293f754019abc3e8599b68378e65dd6b4884ebadc7d607459fb2aR76-R79) [[7]](diffhunk://#diff-e383c735a97293f754019abc3e8599b68378e65dd6b4884ebadc7d607459fb2aR100-R103)

**Participant state management:**

* Enhanced SSE participant state handling in `ConferenceContext` to detect when participants are removed (i.e., missing from SSE event) and mark them as "disconnected" so they can be reconnected if needed, while maintaining accurate status tracking. [[1]](diffhunk://#diff-1f3eb8dfd5a3aeb2a90a00b9ca1d400d6a866915e4df6ce125531477129fc270R123-R129) [[2]](diffhunk://#diff-1f3eb8dfd5a3aeb2a90a00b9ca1d400d6a866915e4df6ce125531477129fc270R194-R228)

### CODE REVIEW

File: teacher-webapp/src/callPage.js [hunk 1]
```diff
@@ -295,6 +297,45 @@ export function DetailsPage() {
...
+    }
+
+    if (!participant || !participant.phoneNumber) {
+      showToast.error("Invalid participant");
+      return;
+    }
+
+    // Confirmation dialog
+    const participantName = participant.name || participant.phoneNumber;
+    const confirmed = window.confirm(
```
Comment (Soumabha) [inline teacher-webapp/src/callPage.js]: window.confirm is a very old way of representing toasts/notifications. We should give better UX by using the Dialog / Toast component instead

File: Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt [hunk 2]
```diff
@@ -604,9 +668,46 @@ class CallViewModel @Inject constructor(
...
+                // Add new participants to tracking set
+                serverParticipantPhones.forEach { phoneNumber ->
+                    participantsInCall.add(phoneNumber)
+                }
+                
+                // Get current state to check for removed participants
+                val currentStudentList = _callState.value ?: emptyList()
+                
+                // Handle participants that are missing from server response (removed)
+                // Keep them as DISCONNECTED so they can be reconnected
```
Comment (Soumabha) [inline Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt]: Duplication of code, move to function and reuse. Use flags for differential handling if any.

**Comment by Soumabha** [comment]
Please add unit tests for the remove functionality.

File: ConferenceV2/app/routers/conference.py [hunk 3]
```diff
@@ -107,7 +107,7 @@ async def add_participant(conference_id: str, phone_number: str):
 
 
 @router.put("/removeparticipant/{conference_id}")
-async def add_participant(conference_id: str, phone_number: str):
+async def remove_participant(conference_id: str, phone_number: str):
```
Comment (Soumabha) [inline ConferenceV2/app/routers/conference.py:110]: Pls verify if it shows up in OpenAPI specs/ Swagger UI.

File: Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt [hunk 4]
```diff
@@ -199,16 +207,30 @@ class CallViewModel @Inject constructor(
...
-                StudentCallStatus(
-                    name = student.name,
-                    phoneNumber = student.phoneNumber,
-                    callerState = CallerState.RINGING,
-                    isMuted = false,
-                    onHold = false,
-                    raiseHand = false
-                )
-            }
+            // Only create initial call statuses for selected students (phoneNumbers)
```
Comment (Soumabha) [inline Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt]: The code logic seems overly complex with multiple variables used for tracking - call status, phones, participants, removed participants.

Instead of having so many variables with individual level tracking, lets create a ParticipantTracker which can exist throughout the call and be referenced at appropirate places to fetch the relevant detail.

```
// dto
data class ParticipantTracker(
    val phoneNumber: String,
    val currentState: CallerState,
    val previousState: CallerState?,
    val hasBeenInCall: Boolean // Replaces participantsInCall set
)

// The latest and greatest map with all details vis-a-vis phone number
private val participantTrackers = mutableMapOf<String, ParticipantTracker>()

```

File: Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt [hunk 5]
```diff
@@ -199,16 +207,30 @@ class CallViewModel @Inject constructor(
...
+            // Filter classroom students to only include those in phoneNumbers
+            val selectedStudentPhones = phoneNumbers.toSet()
+            val initialCallStatuses = args.classroom.students
+                .filter { student -> 
+                    // Normalize phone number for comparison
+                    val normalizedPhone = if (student.phoneNumber.startsWith("91")) {
```
Comment (Soumabha) [inline Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt]: move to a utility class, similar to how we have across web app.

File: ConferenceV2/app/routers/conference.py (hunk 3 above)
Comment (farmanahmed888) [reply ConferenceV2/app/routers/conference.py:110]: image

**Review by a4i-architect — COMMENTED** [review]
### Overall Code Review Summary

---

#### Major Strengths:
- Significant improvements in participant state management, especially in `CallViewModel`, with tracker-based logic.
- Proper separation of concerns between front-end UI (remove participant with confirmation dialog) and back-end API (`removeParticipant` function).
- Focused implementation of participant removal, including optimistic UI updates & eventual consistency.

---

#### Key Issues & Recommendations

**1. API & Frontend Consistency**
- Removing participant is now based on the normalized phone number (adding "91" prefix). Good.
- However, **API URL construction** in `apiService.js` uses query param `phone_number`, but the backend `ConferenceV2/app/routers/conference.py` previously had a route with the path parameter.  
  *Ensure route and API usage match.*  
  **Note:** The code now uses `PUT /removeparticipant/{conference_id}` with query param; confirm backend handles this properly.  
  **Tip:** For API robustness, consider passing `phone_number` in request body instead of URL query.

**2. Front-end Remove UI State**
- `removingIds` array tracks multiple removals; fine.  
- UI disables button during removal; good.
- **Suggestion:** Add user feedback (e.g., toast) on success/failure — which is done.  
- **Nice to have:** Disable the whole participant card or show a spinner overlay for better UX.

**3. ParticipantTracker Logic & Disconnected Handling**
- The logic to handle participants that disappear from server updates by marking them as `DISCONNECTED` is solid, preventing re-adding.  
- **Potential issue:** The code assumes only students can be marked as `DISCONNECTED`. For teachers, the code might not need this.  
- **Tip:** Make role-based distinctions explicit.

**4. Duplicate & Overly Complex Code**
- In `ConferenceContext`, the logic to handle missing participants is duplicated:  
  - In `setParticipantsMap`/`useEffect`, assessing participant removal.  
  - In `computeRemovedParticipantsAsDisconnected`.  
  - **Recommendation:** Consolidate participant status updates into a dedicated function, maybe with explicit role checks.

**5. Type Safety & Code Clarity**
- In `ParticipantTracker`, the fields are clear, but in `CallViewModel`, the logic for `previousState` and `currentState` can be tightened with sealed classes or enums to avoid string misuses.  
- Use Kotlin's `enum class CallerState { CONNECTED, RINGING, DISCONNECTED, UNDEFINED }` for better type safety.

**6. Test Coverage**
- Tests for `ParticipantCard` are comprehensive.
- Tests for API's `removeParticipant` are thorough.
- **But:** No tests (or minimal) for `CallViewModel`'s participant management logic, which is complex.  
  **Suggestion:** Add unit tests verifying tracker state transitions, especially for participants disappearing/reappearing.

**7. Miscellaneous**
- Logging: Use consistent, meaningful logs.
- Comments: Add clearer explanations of the intention behind participant state handling, especially in `computeRemovedParticipantsAsDisconnected`.
- UI Consistency: Confirm removal confirmation dialog is triggered correctly for each participant.

---

### Scoring & Final Remarks

- Code changes address critical participant removal flow well.  
- The approach of using a participant tracker to handle participant state transitions is sound, though it can be refined.

**Score: 78/100**

**Acceptance with Conditions:**
- Confirm API route and payload consistency.
- Simplify or document `computeRemovedParticipantsAsDisconnected` logic.
- Add more unit tests, especially for complex participant state management.
- Consider safer enum types for caller states.

**Optional:** Implement visual cues (e.g., participant disconnection notices) for better UX.  
**Overall:** Good work, just need polishing around API consistency, code clarity, and testing.

**Review by theainerd — APPROVED** [review]
This correctly implements the remove participant functionality. The backend route renaming fixes a bug and the frontend integration is functional.

**Review by Soumabha — APPROVED** [review]

**Review by a4i-architect — COMMENTED** [review]
**Summary of Key Findings & Recommendations**

**Frontend (teacher-webapp):**

- **Participant Removal Flow**:
  - Correctly introduces `onRemove`, `isRemoving` props in ParticipantCard & list components.
  - Adds UX improvements with confirmation dialogs—good.
  - Ensure `isRemoving` disables/removes button reliably; currently, the approach is good but verify state resets to prevent lingering loading states.
  - The code optimizes participant list handling, but manual management of `removeIds` could be simplified by centralizing a "removing" state per participant.

- **Code Quality & Testing**:
  - Tests are comprehensive for remove button behavior.
  - Consider adding tests for failure cases (e.g., API errors during removal).

**Backend (ConferenceV2):**

- **Method Naming & Endpoints**:
  - Correctly renames `add_participant`, `play_audio`, etc.
  - The API call to remove participant uses query params; that's fine, but ensure backend support.

- **Participant State Management**:
  - The participant tracking logic is robust, handling participant transitions, disconnections, and re-connections.
  - Use of `ParticipantTracker` for call state is a good pattern.
  - Implementation of `computeRemovedParticipantsAsDisconnected` is complex but sound; ensure tests cover edge cases.

- **Complexity & Duplication**:
  - Some code segments for tracking participant states are slightly duplicated across polling and socket events; consider centralizing in a helper.
  - The manual handling of participants missing from SSE (mark as disconnected) is necessary but should be verified against backend behavior.

**Design & Architecture:**

- **Participant State Handling**:
  - The explicit `ParticipantTracker` class and `computeRemovedParticipantsAsDisconnected` functions are a good pattern, but complexity can be reduced with more utility functions or state machines.
  - Ensure all edge cases (survivors, reconnections, leftovers) are thoroughly tested.

- **Overall**:
  - The code is leaning towards a complex state machine; consider formalizing state transitions with a dedicated state pattern or enum-based approach.

**Missing / Improvement Suggestions:**

- Validate and unify phone normalization logic across frontend, backend, and utilities.
- Add failure/error handling for API calls in the React components—currently, only success paths are demonstrated.
- Document assumptions regarding call statuses and participant state transitions for maintainability.

**Score**: 78/100  
- The implementation logic is solid, but the complexity and some redundancies reduce clarity. Optimization, better error handling, and more testing coverage could bring it above 85.

---

**References**:
- React state management best practices: https://reactjs.org/docs/hooks-faq.html#how-to-avoid-re-renders
- State machine design patterns: https://refactoring.guru/design-patterns/state
- API error handling strategies: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Error_handling
